Fix #3958: One-off job to use existing stats data to create new ExplorationStatsModels

pranavsid98 opened the pull request:
#3958 

This issue implements the one-off job mentioned in the above issue.

The job also takes into account exploration versioning. Tests have been written for the job as well.

Hi @seanlip PTAL. There are some doc string fixes left - but the functionality is good to go for a review, I believe.
Thanks!

pranavsid98 committed:
create job for migration

pranavsid98 committed:
tests

pranavsid98 committed:
Merge branch 'develop' into migrate_stats_data

pranavsid98 committed:
post merge follow up

pranavsid98 committed:
stats audit job

pranavsid98 committed:
lint fixes

pranavsid98 committed:
small functionality change

pranavsid98 commented:
The doc for the migration can be found at https://docs.google.com/document/d/1u2eUdu0lVKLcV4yiSUWPZWyX9K6nP0Rj3DVH5qiAFgM/edit

pranavsid98 commented:
Hi @LanJosh, could you take a look at this PR and see if everything looks Okay? Thanks!

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
 ])
 
 
+class MigrateStatistics(jobs.BaseMapReduceOneOffJobManager):
```
GenerateV1StatisticsJob

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
 
 
+class MigrateStatistics(jobs.BaseMapReduceOneOffJobManager):
+    """A one-off migration of existing event model instance to generate
```
instance --> instances

pranavsid98 replied:
Done,

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                        'version': str. Version of the exploration.
+                        'state_name': str. Name of the state.
+                        'session_id': str. ID of the session.
+                        'created_on': How many milliseconds ago the event was
```
Shouldn't this be measured using "milliseconds since the Epoch" (or some other fixed point in time, but the Epoch is the default) rather than "how many milliseconds ago"? The latter number changes continuously over time.

pranavsid98 replied:
Yep. The doc string was not right. get_time_millisecs() method returns time since epoch.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                    }
+        """
+        if isinstance(item, stats_models.StartExplorationEventLogEntryModel):
+            yield(
```
Here and elsewhere, I think we normally do "yield (..." (i.e. have a space after 'yield'. You could also format as:

        yield (item.exploration_id, {
            'event_type': feconf.EVENT_TYPE_START_EXPLORATION
        })

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+            yield(item.exploration_id, value)
+
+    @staticmethod
+    def reduce(key, stringified_values):
```
It is fine to just call "key" "exp_id" instead.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        exploration = exp_services.get_exploration_by_id(exp_id)
+        latest_exp_version = exploration.version
+
+        # Check if model already exists. If it does, delete it.
```
Actually, can't you just modify the existing model at the end instead? Two reasons:

(a) If something goes wrong in the code that follows, there'll still be at least some data shown in the UI. The model also only gets updated when the computation below is complete; there isn't this interim phase of "no model available".
(b) A single put() is faster than a delete() + create().

pranavsid98 replied:
Makes sense. Now, it updates the existing model if it exists.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+            exploration_stats_model.delete()
+
+        # The list of state hit events in stringified_values.
+        values_state_hit = []
```
state_hit_event_dicts

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        # The list of state hit events in stringified_values.
+        values_state_hit = []
+        # The list of state answers instances in stringified_values.
+        values_state_answer = []
```
state_answer_event_dicts

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                values_state_answer.append(value)
+
+        state_stats_mapping = {}
+        # Sort list of state hit events in increasing order of it's timestamp.
```
Drop "it's".

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+
+        state_stats_mapping = {}
+        # Sort list of state hit events in increasing order of it's timestamp.
+        values_state_hit = sorted(values_state_hit, key=lambda k: k[
```
Perhaps use 'e' to represent the 'event_dict' in the lambda function.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        values_state_hit = sorted(values_state_hit, key=lambda k: k[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
+            values_state_hit_versioned = [
```
state_hit_events_for_this_version

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        for version in range(1, latest_exp_version+1):
+            values_state_hit_versioned = [
+                val for val in values_state_hit if val['version'] == version]
+            values_state_answer_versioned = [
```
state_answer_events_for_this_version

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        # Create the ExplorationStatsModel instance.
+        exploration_stats = stats_domain.ExplorationStats(
+            exp_id, latest_exp_version, num_starts, 0, num_actual_starts, 0,
+            num_completions, 0, state_stats_mapping)
```
Why are we zeroing out the v2 stuff? If we want to recompute on the v1 in the future then the v2 values should be left as they are.

pranavsid98 replied:
Done. This is applicable only for the case when an ExplorationSttatsModel already exists.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
 
 
+class MigrateStatistics(jobs.BaseMapReduceOneOffJobManager):
+    """A one-off migration of existing event model instance to generate
```
This docstring should make reference to the fact that this job is only supposed to handle v1 statistics, and explain why we're special-casing this.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                val for val in values_state_hit if val['version'] == version]
+            values_state_answer_versioned = [
+                val for val in values_state_answer if val['version'] == version]
+            exploration_versioned = exp_services.get_exploration_by_id(
```
versioned_exploration or exploration

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+            }
+            session_ids_set = set()
+            # Compute total_hit_count and first_hit_count for the states.
+            for value in values_state_hit_versioned:
```
Use a more descriptive name for the loop variable.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                    MigrateStatistics.EVENT_TYPE_STATE_ANSWERS):
+                values_state_answer.append(value)
+
+        state_stats_mapping = {}
```
Add a comment to say that this dict stores cumulative statistics from version 1 to version N of an exploration.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+
+                # Handling state additions, renames and deletions.
+                for change_dict in change_list:
+                    if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
```
One thought: if I have states 1 --> 2 --> 3 --> 4 --> 5 and later on add a state 3a after 3, the state hit counts for 3a will be much lower than those for 4. Is this weird? Are the hit counts still meaningful in that case, from a user's perspective?

pranavsid98 replied:
@seanlip Hmm... It will be weird from the user's POV. But, we should not be assuming values to the state hit counts just to make it meaningful either - this warrants more discussion.

seanlip replied:
Let's leave it as-is for now, I think. We can do something more sophisticated later.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+
+            # Compute num_completions for the states.
+            for session_id in session_ids_set:
+                values_state_hit_sessioned = [
```
This iteration through values_state_hit_versioned multiple times for each session ID is very inefficient and will slow things down a lot. It would be better to figure out a way to do just a single pass through the list and obtain the data keyed by session ID after that single pass.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                values_state_hit_sessioned = [
+                    val for val in values_state_hit_versioned if val[
+                        'session_id'] == session_id]
+                # All states in the above list except the last element (only if
```
I don't understand this comment; please elaborate. What do you mean by "they are counted as completed because they are ordered by their timestamp", and how do you know this is the case?

Also, I don't understand why we ignore the last element if the complete exploration event doesn't exist. What if 4 states out of 7 were hit and the user leaves? The complete exploration event won't exist but all four events represent valid hits, right?

pranavsid98 replied:
@seanlip Let me elaborate more on this: So, lets say there are 5 states in an exploration (1,2,3,4,5). The learner's trace of the session is this: 1, 2, 3, 2, 3, 4 and he quits at state 4. So, now, for each state transition the learner made, there will be a state hit event corresponding to it. So, if we order the state hit events by their creation timestamp, we get the order of states the learner visited. So, for each of the states in the above trace, we increment num_completions by 1. We don't increment num_completions for state 4 because the learner did not progress beyond state 4 and quit there. However, if the final state visited in the learner's trace turns out to be the terminal state, we increment num_completions for all states visited in that trace.

Does that make sense?

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                # are counted as completed because they are ordered by their
+                # timestamp.
+                if session_id not in completed_session_ids:
+                    for value in values_state_hit_sessioned[:-1]:
```
Pull the values_state_hit_sessioned[:-1] out of the if/else. Have the if/else handle only the values_state_hit_sessioned[-1] case.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        num_actual_starts = state_stats_mapping[
+            exploration.states[
+                exploration.init_state_name].interaction.answer_groups[
+                    0].outcome.dest].first_hit_count_v1
```
It is possible for the first answer group to redirect to the same state; you need something more robust here. Check which answer groups lead to a different state, perhaps, and take the sum() or the max() of these. Justify, in comments, why you're calculating this the way you are.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+    @classmethod
+    def require_non_negative(
+            cls, exp_id, property_name, value, state_name=None):
+        if not state_name:
```
Too much duplication:

(a) Pass in just the numerical value of the stat, and the version, separately. So: cls, exp_id, exp_version, property_name, value, state_name=None
(b) Find a way to scope the if/else down, or remove it altogether. Maybe use the second string but have the fourth argument be `... if state_name else ''`.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        stats_services.create_stats_model(exploration_stats)
+
+
+class StatisticsAuditV2(jobs.BaseMapReduceOneOffJobManager):
```
Why is this V2? The audit appears to examine V1 data.

pranavsid98 replied:
Ah. Right. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        """
+        reduced_state_stats_mapping = {
+            state_name: {
+                'total_answers_count': item.state_stats_mapping[state_name][
```
Keep _v1 in the keys. total_answers_count means something different (it means the aggregation of v1 and subsequent versions).

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+        yield(
+            item.exp_id,
+            {
+                'version': item.exp_version,
```
Keep the key as exp_version; be specific.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+            item.exp_id,
+            {
+                'version': item.exp_version,
+                'num_starts': item.num_starts_v1,
```
Keep v1 in the keys.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+            })
+
+    @staticmethod
+    def reduce(key, stringified_values):
```
You can replace key with exp_id.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+
+    @staticmethod
+    def reduce(key, stringified_values):
+        for value in stringified_values:
```
Use a more specific name for the dummy variable.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+                yield ('Completions > starts: exp_id:%s version:%s %s>%s' % (
+                    key, version, num_completions, num_starts),)
+
+            # Number of actual starts must never be less than the number of
```
Some explorations have one state (rare, but possible). How will you handle those?

(I'm not sure if this is the case for _published_ explorations. Might be worth checking whether it's possible to publish a 1-state exploration on the dev server. Also, might be worth prohibiting this going forward; a one-state exploration is unlikely to be of much use.)

pranavsid98 replied:
@seanlip So, one state explorations are possible. We can handle it in this way: Inside the GenerateStatisticsV1Job, our new system ensures that if it does not find any outgoing states from initial state, it sets num_actual_starts to zero. So, in that case, we could set the num_actual_starts to equal num_starts since they have the same meaning.

seanlip replied:
But then such an exploration may change over time to have more than one state, and more generally there may be occasional periods in an exploration's history where it has one state and where it has more than one. Maybe just keeping num_actual_starts at 0 is the consistent thing to do?

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,403 @@
+            # Number of actual starts must never be less than the number of
+            # completions.
+            if num_actual_starts < num_completions:
+                yield ('Completions > actual starts: exp_id:%s version:%s %s>%s' % ( # pylint: disable=line-too-long
```
Consider adding spaces before and after the ">"; ditto below.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off_test.py:
```diff
@@ -0,0 +1,180 @@
+            'TextInput', 0, 0, exp_domain.DEFAULT_OUTCOME_CLASSIFICATION,
+            'session_id2', 0, {}, 'answer2')
+
+        # Update exploration to version 2.
```
Consider moving this to the individual test functions and having separate targeted tests for adding, renaming and deleting. In particular I notice that deletion isn't covered here, and it's also easier to identify issues if there are specific tests. You could also have a fourth test combining the various operations.

pranavsid98 replied:
Added separate tests: 1 test for the first version of the exploration, 1 test for addition & renames, 1 test for deletion. I don't think a separate test for addition might be required since addition and rename tests it sufficiently.

pranavsid98 reviewed (commented):
Hi @seanlip Addressed all comments. Thanks for the detailed pass! PTAL!

pranavsid98 committed:
addressed comments

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,430 @@
+                            'state_name']].useful_feedback_count_v1 += 1
+
+        # Now that the stats for the complete exploration are computed,
+        # calculate num_actual_starts. To compute this, we take the max of the
```
Why max and not sum? Please justify.

pranavsid98 replied:
Added justification in the comments. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,430 @@
+                # All states in the above list except the last element (only if
+                # complete exploration event doesn't exist for the session ID)
+                # are counted as completed because they are ordered by their
+                # timestamp.
```
I might be missing something. At which point are you ordering things by timestamp?

Also, the explanation you gave sounds reasonable. I recommend including more of it in the comment here.

pranavsid98 replied:
So, the state_hit_events_dict is first sorted by order of timestamp. Then, we compute state_hit_events_for_this_version, which is also sorted in order of timestamp. Then, for each state hit event in the above list, we add it to a dict keyed by its sessionID (state_hit_events_session_id_mapping). So, for each sessionID in the above dict, all corresponding state hit events are ordered by their timestamp.

Re including more details in the comment, Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,430 @@
+        state_hit_event_dicts = sorted(state_hit_event_dicts, key=lambda e: e[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
+            print version
```
Drop this, please.

pranavsid98 replied:
Done.

pranavsid98 committed:
changes

pranavsid98 commented:
Hi @seanlip . For some reason, your comment regarding the one state explorations vanished, I still have it on email though.

So, fundamentally, we can't set num_actual_starts to zero for one state explorations because if somebody is completing an exploration, it makes sense for them to have actually started it. I'd still prefer setting num_actual_starts to num_starts in this case.

If the exploration is going to have only one state at some point of time, then the learner will receive correct stats at that point. When he adds more states to it, now num_actual_starts will start becoming more sparse and gain more meaning. I think this might be the right thing to do.

What do you think?

seanlip commented:
That's a good point re consistency with num_completions. I think what you propose is fine. To make it simpler going forward we might want to limit published explorations to have at least two states so that the previous discussion is a consideration only for v1 stats. That should be simple to do and I think it also makes sense from a pedagogical perspective.

Also, btw, I just realized you were handling add/delete/rename state but it's unclear to me how you are planning to handle reverts?

pranavsid98 committed:
exp reverts

pranavsid98 commented:
Hi @seanlip . I've added functionality to handle exploration reverts and also added a test for that. I think this PR is up for another review. PTAL!

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+        # Sort list of state hit events in increasing order of timestamp.
+        state_hit_event_dicts = sorted(state_hit_event_dicts, key=lambda e: e[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
```
nit: add spaces around "+"

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+        state_hit_event_dicts = sorted(state_hit_event_dicts, key=lambda e: e[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
+            start_exploration_events_for_this_version = [
```
This is a nested for loop. Don't do this. Group by version at the outset. Otherwise your code is going to time out in production (we have millions of events).

Ditto below.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            state_answer_events_for_this_version = [
+                val for val in state_answer_event_dicts if val[
+                    'version'] == version]
+            versioned_exploration = exp_services.get_exploration_by_id(
```
If you are going to end up using all the versions of the exploration then you might as well do a get_multi() at the outset.

pranavsid98 replied:
@seanlip Can we do that? I've been looking up base models and ndb only supports get_multi for different IDs. We could get a snapshot reconstituted set of all version instances, but is that required?

seanlip replied:
I don't see what the issue is with different IDs, sorry. Can you elaborate? There's an ndb get() buried somewhere in the current flow and it's possible you might need to add a new function in the correct place, perhaps in VersionedModel or exp_services.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            versioned_exploration = exp_services.get_exploration_by_id(
+                exp_id, version=version)
+
+            is_revert = False
```
You might as well just store `revert_to_version = None` here. If it ends up being non-None then it's a revert.

pranavsid98 replied:
Yep. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+                                change_dict['old_state_name']))
+                    elif change_dict['cmd'] == 'AUTO_revert_version_number':
+                        revert_to_version = change_dict['version_number']
+                        is_revert = True
```
See above regarding redundancy of this variable.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+                        is_revert = True
+
+            # Computing num_starts and num_completions.
+            num_starts += len(start_exploration_events_for_this_version)
```
If this is the only thing you're going to take away from the event dict, then maybe just keep track of only the counter at the outset? No need to make a list of all "start event" dicts, etc.

Similar comments apply for other fields. If you can save space, try to do so. And the fewer times you iterate through the entire dataset, the better, so try and pick a data structure that allows you to easily update as much stuff as possible on-the-fly in a single pass through all the data at the outset.

pranavsid98 replied:
Done. Found multiple places where we could reduce iterating through the dataset.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            # Dict mapping each state name of the exploration and the list of
+            # unique session IDs that have entered that state.
+            state_session_ids = {
+                state_name: [] for state_name in versioned_exploration.states
```
Consider using a set() as the value.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            }
+            session_ids_set = set()
+            # Compute total_hit_count and first_hit_count for the states.
+            for state_hit_event in state_hit_events_for_this_version:
```
Again, nested for loop; seems wasteful. Try to avoid.

pranavsid98 replied:
@seanlip I think this nested for loop will be necessaery. For this particular nested loop, it os O(n) complexity where n is the number of total state_hit_events. So, it is still technically only one pass through the state hit event dataset.

seanlip replied:
No, O(n) only measures asymptotic complexity. Two O(n) passes are still two passes; getting it down to one reduces the time spent by a factor of 2 (approximately; there's going to be some overhead too).

In this case, could you not compute the information on first pass through the state_hit_event list for all versions? E.g. store counts/session IDs in a nested dict with successive version number and state name keys? This might mean you'll need to get all state names for all versions of the exploration at the outset, but that's also reasonable since you're processing all versions in this step anyway.

pranavsid98 replied:
Done.

seanlip reviewed (commented):
(Haven't finished reading, but I need to run. I'll leave this here for now and pick up again later.)

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            # Compute total_answers_count and useful_feedback_count.
+            for value in state_answer_events_for_this_version:
+                for answer in value['submitted_answer_list']:
+                    state_stats_mapping[value[
```
Worth just using len(value['submitted_answer_list']) for this? I think len() is an O(1) operation in Python.

In fact if you only need the counts then why not pass only this info over from the map() stage? Answers are big.

pranavsid98 replied:
That makes sense. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            # the first hit counts of all states that the initial state leads
+            # to. The max is taken instead of sum because the sum of
+            # first_hit_counts will sometimes end up counting a single learner
+            # twice. The max of both first hit counts ensures that we are
```
Well, I guess max could undercount... But I'm fine with this.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            # counting only unique learners who traversed past the initial
+            # state.
+            max_first_hit_from_init_state = 0
+            for answer_group in versioned_exploration.states[versioned_exploration.init_state_name].interaction.answer_groups: # pylint: disable=line-too-long
```
Pull versioned_exploration.states[versioned_exploration.init_state_name] into a separate variable named init_state. In general try to avoid pylint pragmas where possible.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+                    if state_stats_mapping[
+                            answer_group.outcome.dest].first_hit_count_v1 > (
+                                max_first_hit_from_init_state):
+                        max_first_hit_from_init_state = state_stats_mapping[
```
Mm, can just do max_first_hit_from_init_state = max([answer_group for answer_group in ... if ...])

pranavsid98 replied:
Done. Getting better at dict comprehensions :)

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+                                max_first_hit_from_init_state):
+                        max_first_hit_from_init_state = state_stats_mapping[
+                            answer_group.outcome.dest].first_hit_count_v1
+            if max_first_hit_from_init_state == 0:
```
Please add comment explaining what's going on here. Also, I think this is the wrong condition; you should check the number of states explicitly.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+                num_actual_starts = old_exploration_stats.num_actual_starts_v1
+                state_stats_mapping = old_exploration_stats.state_stats_mapping
+            if exploration_stats_model is not None:
+                exploration_stats = stats_domain.ExplorationStats(
```
Maybe store up these domain objects and do a bulk put_multi() at the end? Some explorations have 100s of versions; a put_multi() will be much faster than 100+ individual puts.

pranavsid98 replied:
Sounds right. We needed to store stats models for retrieval during reverts, but we can work around that by retrieving directly from out list of domain objects. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+                old_exploration_stats = (
+                    stats_services.get_exploration_stats_by_id(
+                        exp_id, revert_to_version))
+                num_starts = old_exploration_stats.num_starts_v1
```
Er, wait a sec. If we're just going to throw away all the data above (in this case) then why compute it in the first place?

Consider extracting some of the above into subsidiary functions and calling them only if needed. Then you can do: `if is_revert: ... else: ...` more cleanly here, where the subsidiary functions are only called in the else branch.

pranavsid98 replied:
@seanlip Ahh. There is a small catch here. Throwing away the data above is not right. Lets say we have this use case: v1->v2->v3->v4 and v4 is a revert to v1. Now, we have aggregated stats data for v1, v2 and v3. Inside the for loop, while handling the events for v4, they need to be taken into account. But, aggregation should not be added to the values accumulated until v3, rather they should be added to the events accumulated until v1. So, we can't throw away this data. Rather, we should be setting all values of stats to the values of stats for the version we reverted to and continue accumulating from there.
I think this is how it needs to be done. Does this sound right?

seanlip replied:
Yes, that sounds reasonable.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,466 @@
+            # Number of actual starts must never be less than the number of
+            # completions.
+            if num_actual_starts_v1 < num_completions_v1:
+                yield ('Completions > actual starts: exp_id:%s version:%s %s > %s' % ( # pylint: disable=line-too-long
```
Refrain from pragmas, here and below. Instead try something like:

    yield (
        'Completions > ...'
        '...' % (
            exp_id, exp_version, ...),)

pranavsid98 replied:
Done.

seanlip reviewed (changes requested):
OK, this finishes a full pass! I'm mostly concerned about scaling issues, so please try and see if you can make the job as time/space-efficient as possible.

pranavsid98 committed:
changes

pranavsid98 reviewed (commented):
@seanlip Thanks for the extensive review. Cleaned up stuff throughout and we have significant progress. Executing the tests used to take 91 seconds whereas I takes only 62 seconds now. PTAL! I think it looks nice now.

pranavsid98 committed:
added get multi

pranavsid98 commented:
Hi @seanlip Addressed all review comments. Specifically, added get_multi for the explorations and also reduced almost all of the nested for loops. All computation takes place in the initial pass through the dataset (btw, this wasn't possible for num_completions alone since our way of computing this involves sorting the state hit events in order of their timestamp. This cannot be done in the first pass, therefore, this computation alone involves a second pass through the state hit events). PTAL!
Thanks

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+                exp_id, version_numbers))
+
+        # The list of exploration start counts mapped by version.
+        start_exploration_counts_by_version = {}
```
Use collections.defaultdict here and below (where relevant), and then you don't need all the if conditions to check whether or not the key already exists. Currently there are far too many if conditions, which makes the code more complicated and makes it harder to keep track of the logic.

pranavsid98 replied:
@seanlip Wow. New stuff in Python I learn each day. This data structure is beautiful and reduced so many lines of code. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+        completed_session_ids = []
+        # Dict mapping versions -> sessionID -> StateHitEvents.
+        session_id_state_hit_event_mapping = {}
+        # Dict mapping each state name of the exploration and the list of
```
and --> to

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+        # Dict mapping versions -> sessionID -> StateHitEvents.
+        session_id_state_hit_event_mapping = {}
+        # Dict mapping each state name of the exploration and the list of
+        # unique session IDs that have entered that state.
```
You're initializing the values below with dicts, which don't look like lists to me. Why not sets, though?

pranavsid98 replied:
@seanlip Doc string is slightly off. This is a mapping from versions -> state_name -> set(sessionIDs)

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+            value = ast.literal_eval(value_str)
+
+            if value['event_type'] == feconf.EVENT_TYPE_START_EXPLORATION:
+                if str(value['version']) not in (
```
Is it actually necessary to convert the version number to a string?

pranavsid98 replied:
@seanlip Small pointer I remembered from the ML project, not to keep dict keys as integers. I think integer keys might still work though. But, I just have this programming practice of not using integer keys. Is that fine?

seanlip replied:
I thought that was only for JS. For something like this which is completely in the backend and has no chance of making it to the frontend, int keys should be fine.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+        # The dict of state answer counts mapped by version.
+        state_answer_counts_by_version = {}
+        # The list of session_ids of learners completing an exploration.
+        completed_session_ids = []
```
Worth using set() and set.add() (to avoid duplicates)?

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+        state_session_ids_by_version = {}
+
+        for value_str in stringified_values:
+            value = ast.literal_eval(value_str)
```
It may not be a bad idea to define `version = value['version']` here. It makes many of the following lines shorter and less cumbersome.

The same may be true with session_id in other parts of this code.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+                        value['state_name']] = set()
+                state_hit_counts_by_version[str(value['version'])][value[
+                    'state_name']]['total_hit_count'] += 1
+                if value['session_id'] not in state_session_ids_by_version[
```
This is effectively an inner for loop, which should be avoided. Why not just add the session IDs to a set and then, at the very end, take the length of the set?

pranavsid98 replied:
Makes sense. List lookups are O(n). Changed this.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+        exploration_stats_list = []
+        for version in version_numbers:
+            state_hit_counts_for_this_version = {}
+            if str(version) in state_hit_counts_by_version:
```
If you use collections.defaultdict then you shouldn't need this and other if clauses.

pranavsid98 replied:
Yep. Removed.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+                # So, if the last state is not the terminal state, we increment
+                # num_completions for all states except the last state.
+                # Otherwise, we increment num_completions for all the states.
+                for state_hit_event in state_hit_event_dicts_sessioned[:-1]:
```
I don't think I understand why this can't be optimized away. If you create a dict keyed by version whose values are dicts keyed by state name, couldn't you just increment the inner values when walking along the state hit event dicts the first time?

Also, separately, I guess you want to keep track of the event with the largest timestamp. Why not do that in a dict keyed by version and session ID? The value of the dict can be the state name containing the largest timestamp found so far. When iterating along the full list of events the first time, you can just update that value if needed.

pranavsid98 replied:
@seanlip I've got a small issue with this idea. So, our plan is to increment num_completions for all state hit events within a session except the last one (that is where the learner quit).
Now, in what order are events present in values_stringified? When we are doing the first pass over all values, we don't really have any ordering to the events in it right? So, any state hit event we encounter, we won't really know if it is the state hit event with the largest timestamp, would we (since there could be a later event with a larger timestamp)? That was why I thought we had to process all events once, and then sort them again to compute num_completions correctly.

An alternative could be to just increment num_completions for all the state hit events we encounter the first time while maintaining a data structure to keep track of the largest timestamp keyed by session. Then, we can just run a pass through all sessions to find that state hit event with largest timestamp and decrement num_completions for all those values. Would this work?

seanlip replied:
It's true that initially we don't have any ordering. But all you want to do with the ordering is to keep track of which event has the largest timestamp, and the method I proposed accounts for that. I.e. you keep track of the state name corresponding to the largest timestamp on a rolling basis as you make your first pass through the list, updating it as you get more info.

It's a bit like finding the max value in a list by iterating through the list and storing the current max value seen so far, and updating that stored variable only when you find a larger value.

pranavsid98 replied:
@seanlip Yep. I agree that we can find the largest timestamp event on a rolling basis on the first pass. But, from that point, computing num_completions will still be the same complexity. Since our data structure is a dict keyed by version its value is a dict keyed by sessionID, we'll have to iterate over all sessionIDs for a version to increment num_completions anyway. Do I seem to be missing something?

pranavsid98 replied:
@seanlip So, I've figured another way out of this. Instead of doing any of these computations inside the for loop, I did it at the outset. This still involves another pass over the data, but this is just one extra pass and nothing inside the nested for loop. How does it look now?

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+            max_first_hit_from_init_state = max([
+                state_stats_mapping[
+                    answer_group.outcome.dest].first_hit_count_v1
+                for answer_group in init_state.interaction.answer_groups if (
```
Indent by 4. Move "if" part to next line, still indented by 4.

pranavsid98 replied:
That doesn't work? The 'for' needs to be aligned with state_stats_mapping (on the same indent level).

seanlip replied:
Sorry! This one is my error. I misread.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+            if len(versioned_exploration.states) == 1:
+                num_actual_starts = num_starts
+            else:
+                num_actual_starts = max_first_hit_from_init_state
```
You might as well put the above calculation in this branch so that it's not done unnecessarily?

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+
+            # Check if model already exists. If it does, update it, otherwise
+            # create a fresh ExplorationStatsModel instance.
+            exploration_stats_model = (
```
Do get_multi() at outset rather than incremental get() operations here.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -18,14 +18,541 @@
+            exploration_stats_list.append(exploration_stats.to_dict())
+
+        # Save all the ExplorationStats models to storage.
+        stats_models.ExplorationStatsModel.create_multi(exploration_stats_list)
```
put_multi()? create_multi() assumes nonexistence.

pranavsid98 replied:
I did use put_multi in the storage layer. Naming quirk, fixed it.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -530,6 +530,28 @@ def get_version(cls, entity_id, version_number):
+        instances = []
+        # pylint: disable=protected-access
+        for version in version_numbers:
+            cls.get(entity_id)._require_not_marked_deleted()
```
Why do this check here? The deleted status would be the same for all versions, right -- so we can do the check at the outset?

pranavsid98 replied:
Yep. Done. Checking at the outset now.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -132,53 +140,113 @@ def reduce(exp_id, stringified_values):
         exploration = exp_services.get_exploration_by_id(exp_id)
```
Might want to check the case where an exploration is deleted.

pranavsid98 replied:
Done.

pranavsid98 committed:
changes addressed

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,515 @@
+                exp_id, version_numbers))
+
+        # The list of exploration start counts mapped by version.
+        start_exploration_counts_by_version = collections.defaultdict(lambda: 0)
```
FYI I've generally seen `collections.defaultdict(int)` for this common case; you can save lambda for the harder stuff. Just a thought.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,515 @@
+                'first_hit_count': 0
+            }))
+        # The dict of state answer counts mapped by version.
+        state_answer_counts_by_version = collections.defaultdict(lambda: {})
```
Just checking -- you don't want the default value here to be a defaultdict too?

pranavsid98 replied:
Hmm.. Yes, actually. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,515 @@
+        exploration_stats_list = []
+        for version in version_numbers:
+            state_hit_counts_for_this_version = {}
+            state_hit_counts_for_this_version = (
```
Why are you defining this variable twice?

pranavsid98 replied:
Missed to remove that. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,515 @@
+            state_answer_counts_for_this_version = (
+                state_answer_counts_by_version[str(version)])
+            session_id_state_hit_event_mapping_for_this_version = {}
+            session_id_state_hit_event_mapping_for_this_version = (
```
Why are you defining this variable twice?

pranavsid98 replied:
Removed

pranavsid98 committed:
fixes

pranavsid98 committed:
Merge branch 'develop' into migrate_stats_data

pranavsid98 committed:
lint fix

pranavsid98 committed:
completion counts

pranavsid98 committed:
changes

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+        start_exploration_counts_by_version = collections.defaultdict(int)
+        # The list of exploration complete counts mapped by version.
+        complete_exploration_counts_by_version = collections.defaultdict(
+            lambda: 0)
```
Why not defaultdict(int)?

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+            }))
+        # The dict of state answer counts mapped by version.
+        state_answer_counts_by_version = collections.defaultdict(
+            lambda: collections.defaultdict(lambda: {}))
```
I'm not sure I understand this; why is the inner `lambda: {}` not just `int`?

pranavsid98 replied:
@seanlip Because the inner state_answers_counts are stored in a dict with two keys - total_answers_count and useful_feedback_count.

seanlip replied:
As discussed, specify these keys explicitly and aggregate, similar to state_hit_counts_by_version.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+        state_answer_counts_by_version = collections.defaultdict(
+            lambda: collections.defaultdict(lambda: {}))
+        # The set of session_ids of learners completing an exploration.
+        completed_session_ids = set()
```
Maybe call this exp_completed_session_ids since there's also the notion of state completion.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+                    session_id_largest_timestamp_mapping[version][
+                        session_id] = value
+                session_id_state_hit_event_mapping[version][
+                    session_id].append(value)
```
The comment above for this dict suggests that each entry is a list of state names; why are you appending the event dict?

pranavsid98 replied:
The doc string was off. I was initially storing state names, but that information wasn't enough, so had to store the events.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+            elif value['event_type'] == (
+                    GenerateV1StatisticsJob.EVENT_TYPE_STATE_ANSWERS):
+                state_answer_counts_by_version[version][
+                    value['state_name']] = {
```
Are state names unique in these models? If not, should you be adding the values rather than overwriting?

pranavsid98 replied:
@seanlip Yep. state names are unique in this case. For one state, we'll always have only one StateAnswersModel.

seanlip replied:
Per discussion, it turns out that they aren't unique; this needs to be addressed.

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+            lambda: collections.defaultdict(int))
+        # The dict mapping version -> sessionID -> StateHitEvent with largest
+        # timestamp.
+        session_id_largest_timestamp_mapping = collections.defaultdict(
```
If you're storing an actual event, should this be called session_id_latest_event_mapping?

pranavsid98 replied:
Sounds good. Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+            lambda: collections.defaultdict(lambda: {}))
+        # The set of session_ids of learners completing an exploration.
+        completed_session_ids = set()
+        # Dict mapping versions -> sessionID -> state_names.
```
From reading the code below, isn't the target value an event (rather than a state name)?

pranavsid98 replied:
Changed.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+                # leads to. The max is taken instead of sum because the sum of
+                # first_hit_counts will sometimes end up counting a single
+                # learner twice. The max of both first hit counts ensures that
+                # we arecounting only unique learners who traversed past the
```
arecounting --> are counting

pranavsid98 replied:
Done.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+            # create a fresh ExplorationStatsModel instance.
+            exploration_stats = exploration_stats_by_version[version - 1]
+            if exploration_stats is not None:
+                exploration_stats = stats_domain.ExplorationStats(
```
Do you need to store this back in the exploration_stats_by_version list (in order to handle later reverts)?

pranavsid98 replied:
@seanlip I don't think so. We do store it in the exploration_stats_list which is used for handling reverts.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -16,16 +16,524 @@
+                exploration_stats = stats_domain.ExplorationStats(
+                    exp_id, version, num_starts, 0, num_actual_starts, 0,
+                    num_completions, 0, state_stats_mapping)
+            exploration_stats_list.append(exploration_stats.to_dict())
```
Can you pick names that distinguish this more clearly from exploration_stats_by_version?

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -530,6 +530,27 @@ def get_version(cls, entity_id, version_number):
+            version_numbers: list(int). List of version numbers.
+
+        Returns:
+            list(VersionedModel|None). Model instances representing the given
```
In what case will a None value be returned here?

pranavsid98 replied:
@seanlip Nice catch. This never returns None. We check that the exploration has not been deleted before we run this and it returns list(VersionedModel) always. Added a doc string to mention that.

pranavsid98 replied:
Oh wait, it does return None when the model does not exist for a particular version. This ca happen for the exploration stats model. It's better to be kept this way.

seanlip replied:
Per discussion, please investigate and ensure the function is robustly-defined in all cases. Add tests as needed.

pranavsid98 replied:
Done.

seanlip reviewed (commented):
This looks very good! Remaining comments are minor. Thanks @pranavsid98!

pranavsid98 committed:
addressed minor review comments

pranavsid98 committed:
changes

pranavsid98 commented:
@seanlip Changes made! PTAL!

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -538,10 +540,14 @@ def get_multi_versions(cls, entity_id, version_numbers):
                 versions.
         """
         instances = []
+        try:
```
This is not robust. What if the entity with the given id exists but at least one of the version numbers is higher than the current version?

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -538,10 +540,14 @@ def get_multi_versions(cls, entity_id, version_numbers):
+        try:
+            cls.get(entity_id)
+        except Exception as  e:
+            logging.error(e)
```
Why catch and log? If this is not expected to happen, raise directly.

In designing this particular API, you'll need to figure out: (a) is the client expected to ensure that the exp id and version numbers are valid, or (b) is this method supposed to take care of that? If (a), specify that constraint clearly in the docstring. If (b), specify what is returned/raised in cases of error, and add tests.

When you have a definition, just throw exceptions for stuff that's not supposed to happen in practice (no need to catch them). Otherwise, handle exceptional cases as described in the docstring. The docstring should fully and clearly describe what should happen for each different type of input.

pranavsid98 replied:
Done.

seanlip commented on core/domain/exp_services.py:
```diff
@@ -256,17 +256,19 @@ def get_exploration_summary_by_id(exploration_id):
+            None is returned, the exploration does not exist.
     """
     explorations = []
+    if not get_exploration_by_id(exp_id):
```
Is this meant to be for error-checking? If so, please see my comments in base_model/gae_models.py -- they apply here, too.

pranavsid98 replied:
Done.

seanlip commented on core/domain/exp_services.py:
```diff
@@ -256,17 +256,19 @@ def get_exploration_summary_by_id(exploration_id):
 
     Returns:
-        list(Exploration). List of Exploration domain class instances.
+        list(Exploration)|None. List of Exploration domain class instances. If
```
I'm not sure about this. It's not in line with how our other "get multi" calls work, and it doesn't take into account the case where some version numbers might be invalid...

pranavsid98 replied:
Done.

seanlip reviewed (commented):
Just a general note: I don't see added tests for the invalid exp_id case, or invalid version number case, or for the case of multiple StateAnswersModel, or for other cases that you want to preemptively catch in the MR job. These are important, so please make sure to add them.

pranavsid98 commented:
@seanlip. I've added tests for invalid exp_id, version numbers inside the exp_services_test module for the get_multiple_explorations_by_version method. But, we shouldn't be encountering fake IDs/ version numbers in the job because it only iterates through existing ExplorationModels and uses their IDs. I don't think fake IDs might creep up inside the job (they are handled all the same).

Regarding the multiple StateAnswersModel testing issue, I don't think we can explicitly create shared versions of StateAnswersModel with different shard IDs. The insert_submitted_answers method takes care of that for us. Do you have any suggestions regarding this?

pranavsid98 committed:
changes addressed

seanlip commented on core/domain/exp_services_test.py:
```diff
@@ -323,6 +323,56 @@ def test_retrieval_of_explorations(self):
         with self.assertRaises(Exception):
             exp_services.get_exploration_by_id('fake_exploration')
 
+    def test_retrieval_of_multiple_exploration_versions(self):
```
Break this into separate tests: one for fake exp ID, one for correct behavior of versions, one for version number errors.

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -542,17 +540,34 @@ def get_multi_versions(cls, entity_id, version_numbers):
                 versions.
+
+        Raises:
+            ValueError. Given entity_id is invalid.
```
Given --> The given

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -542,17 +540,34 @@ def get_multi_versions(cls, entity_id, version_numbers):
+            ValueError. Given entity_id is invalid.
+            ValueError. Requested version number cannot be higher than the
+                current version number.
+            ValueError. Version number is invalid.
```
At least one version number is invalid.

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -542,17 +540,34 @@ def get_multi_versions(cls, entity_id, version_numbers):
+
+        entity = cls.get(entity_id, strict=False)
+        if not entity:
+            raise ValueError('Given entity_id %s is invalid.' % (entity_id))
```
Given --> The given

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -542,17 +540,34 @@ def get_multi_versions(cls, entity_id, version_numbers):
+        if not entity:
+            raise ValueError('Given entity_id %s is invalid.' % (entity_id))
+        current_version = entity.version
+        if version_numbers[-1] > current_version:
```
Do not assume that the version numbers are sorted.

pranavsid98 replied:
Done. Using max() instead now.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -542,17 +540,34 @@ def get_multi_versions(cls, entity_id, version_numbers):
-            instances.append(cls(id=entity_id)._reconstitute_from_snapshot_id(
-                snapshot_id))
+            try:
+                entity = cls(id=entity_id)._reconstitute_from_snapshot_id(
```
I just realized that this is basically a get() in a for loop. I think you'll need a "multi" version of reconstitute as well (or maybe you can extract the get part from it, do the get outside reconstitute(), and pass the data in).

pranavsid98 replied:
Done.

pranavsid98 committed:
changes

pranavsid98 commented:
Hi @seanlip
I've added tests for large state answers in the job_test now. Also, the get_multi_Versions is modified such that it does a get_multi at the outset to retrieve all snapshot content classes and then populates the instance in a for loop.

Oh, and re writing tests for VersionedModel, the commit() also doesn't work because we don't have any defined snapshot content/ snapshot metadata class for the base versioned model (they are initialised to None in the model). These classes are defined by the models that inherit the base versioned model. So, while testing, we can't test VersionedModel straight up. If testing is required, we'll probably have to create a class, make it inherit VersionedModel and then test that.

seanlip commented:
Creating a dummy test class that inherits VersionedModel sounds absolutely fine to me, and was actually the way I had envisioned testing it!

seanlip commented on core/domain/exp_services.py:
```diff
@@ -263,12 +263,9 @@ def get_multiple_explorations_by_version(exp_id, version_numbers):
     Returns:
-        list(Exploration)|None. List of Exploration domain class instances. If
-            None is returned, the exploration does not exist.
+        list(Exploration). List of Exploration domain class instances.
```
List of Exploration domain objects.

pranavsid98 replied:
Done.

seanlip commented on core/domain/exp_services_test.py:
```diff
@@ -323,6 +323,77 @@ def test_retrieval_of_explorations(self):
+        exploration = self.save_new_default_exploration(
+            self.EXP_ID, self.owner_id)
+
+        # Update exploration to version 2.
```
No need to change this, but as a general rule, try to keep tests minimal. In this case you don't really need the subsequent versions (or, at worst, you just need version 2) in order to test what you're testing here.

seanlip commented on core/domain/stats_jobs_one_off_test.py:
```diff
@@ -122,6 +123,48 @@ def test_creation_of_stats_model_for_v1(self):
+
+        self.assertEqual(
+            exploration_stats.state_stats_mapping[
+                'Home'].total_answers_count_v1, 602)
```
This seems slightly odd to me. I understand the 600 and the 400 but where do the extra 2 and 1 come from?

It's probably legit, but please add a comment to clarify.

pranavsid98 replied:
@seanlip Oh, the 2 is from the couple of answers in the setup method

seanlip replied:
OK, please add a comment to clarify.

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models_test.py:
```diff
@@ -17,6 +17,8 @@
 from core.platform import models
 from core.tests import test_utils
 
+import feconf
```
Why add this?

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -542,17 +540,42 @@ def get_multi_versions(cls, entity_id, version_numbers):
+        current_version = entity.version
+        if max(version_numbers) > current_version:
+            raise ValueError(
+                'Requested version number cannot be higher than the current '
```
Consider adding a bit more info, e.g. 'Requested version number %s cannot ... current version number %s'. Helps with debugging.

pranavsid98 replied:
Done.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -542,17 +540,42 @@ def get_multi_versions(cls, entity_id, version_numbers):
+            snapshot_ids.append(snapshot_id)
+
+        snapshot_models = cls.SNAPSHOT_CONTENT_CLASS.get_multi(snapshot_ids)
+        for index, snapshot_model in enumerate(snapshot_models):
```
Why do you need index/enumerate? I don't think you use it.

pranavsid98 replied:
Removed it.

seanlip reviewed (commented):
Thanks, @pranavsid98! The change to the "reconstitute" functionality is good; it still needs tests (see previous comment).

pranavsid98 committed:
changes

pranavsid98 commented:
Hi @seanlip PTAL! I've added tests for VersionedModel now.

seanlip commented on core/domain/exp_services_test.py:
```diff
@@ -382,15 +380,15 @@ def test_version_number_errors_for_get_multiple_exploration_versions(self):
-                'version number.'):
+            ValueError,
+            'Requested version number %s cannot be higher than the current '
+            'version number %s.' % (4, 3)):
```
In tests, just hardcode the expected message directly. Don't interpolate constants.

pranavsid98 replied:
Done.

seanlip reviewed (approved):
Two small nits, otherwise LGTM. Thanks!

pranavsid98 committed:
final nit changes
